File: carbondesign/mithril-js/tooltip.js
import m from 'mithril/hyperscript';
//-
import { Node, modify_svg } from './base';

export class Tooltip extends Node
{
    WANT_CHILDREN = true
    SLOTS = ['icon', 'footer']
    MODES = ['interactive', 'definition', 'icon']
    NODE_PROPS = ['align', 'position']

    prepare(vnode, values, context)
    {
        if (vnode.attrs.align)
        {
            values['class'].push(`bx--tooltip--align-${vnode.attrs.align}`);
        }
        if (vnode.attrs.position)
        {
            values['class'].push(`bx--tooltip--${vnode.attrs.position}`);
        }
    }

    render_interactive(vnode, values, context)
    {
        return (
//##
m.fragment(null,
  [
    m('div',
      {
        id: `label-${values.id}`,
        'class': `bx--tooltip__label ${values['label_class']}`,
        ...values.props,
      },
      [
          values.label,
          m('button',
            {
              'aria-expanded': false,
              'aria-labelledby': `label-${values.id}`,
              'data-tooltip-trigger': '',
              'data-tooltip-target': `#${values.id}`,
              'class': `bx--tooltip__trigger ${values['class']}`,
              'aria-controls': values.id,
            },
            this.slot('icon', vnode, values, context)),
      ]),
    m('div.bx--tooltip',
      {
        id: values.id,
        'aria-hidden': true,
        'data-floating-menu-direction': 'bottom',
      },
      [
        m('span.bx--tooltip__caret'),
        m('div.bx--tooltip__content',
          {
            tabindex: -1,
            role: 'dialog',
            'aria-describedby': `body-${values.id}`,
            'aria-labelledby': `label-${values.id}`,
          },
          [
            m('div', {id: `body-${values.id}`}, values.child),
            this.slot('footer', vnode, values, context)),
          ]),
        m('span', {tabindex: 0}),
      ]),
  ])
//##
        );
    }

    render_definition(vnode, values, context)
    {
        return (
//##
m('div.bx--tooltip--definition.bx--tooltip--a11y',
  {
    'data-tooltip-definition': '',
  },
  [
    m('button',
      {
        'aria-describedby': values.id,
        'class': `bx--tooltip__trigger bx--tooltip--a11y bx--tooltip__trigger--definition ${values['class']}`,
      },
      values.label),
    m('div.bx--assistive-text',
      {
        id: values.id,
        role: 'tooltip',
      },
      values.child),
  ])
//##
        );
    }

    render_icon(vnode, values, context)
    {
        return (
//##
m('button',
  {
    'class': `bx--tooltip__trigger bx--tooltip--a11y ${values['class']}`,
    'data-tooltip-icon': '',
  },
  [
    m('span.bx--assistive-text', null, values.child),
    this.slot('icon', vnode, values, context),
  ])
//##
        ); 
    }

    render_slot_icon(vnode, values, context)
    {
        return modify_svg(values.child, {
            focusable: false,
            preserveAspectRatio: 'xMidYMid meet',
            fill: 'currentColor',
            style: 'width:16px; height:16px',
            'aria-hidden': true,
        })
    }

    render_slot_footer(vnode, values, context)
    {
        return (
//##
m('div',
  {
    'class': `bx--tooltip__footer ${values['class']}`,
    ...values.props,
  },
  values.child)
//##
        );
}


export class TooltipHeading extends Node
{
    WANT_CHILDREN = true
    DEFAULT_TAG = 'h4'

    render_default(vnode, values, context)
    {
        return (
//##
m(values.tag,
  {
    'class': `bx--tooltip__heading ${values['class']}`,
    ...values.props,
  },
  values.child)
//##
        );
    }
}

File: carbondesign/tags/tooltip.py
"""
Tooltip
=======

See: https://www.carbondesignsystem.com/components/tooltip/usage/

Tooltips display additional information upon click, hover, or focus. The
information should be contextual, useful, and nonessential.

Overview
--------

A tooltip is a message box that is displayed when a user hovers over, clicks
or gives focus to a UI element such as an icon, a highlighted word, or a button.
""" # pylint:disable=line-too-long
# pylint:disable=too-many-lines

from .base import Node, modify_svg

class BaseTooltip(Node):
    """Base tooltip.
    """
    WANT_CHILDREN = True
    "Template Tag needs closing end tag."
    NODE_PROPS = ('id', 'align', 'position')
    "Extended Template Tag arguments."
    POSSIBLE_ALIGN = ('start', 'center', 'end')
    "Documentation only."
    POSSIBLE_POSITION = ('top', 'right', 'bottom', 'left')
    "Documentation only."

    def prepare(self, values, context):
        """Prepare values for rendering the templates.
        """
        align = self.eval(self.kwargs.get('align'), context)
        if align and align in self.POSSIBLE_ALIGN:
            values['class'].append(f'bx--tooltip--align-{align}')

        position = self.eval(self.kwargs.get('position'), context)
        if position and position in self.POSSIBLE_POSITION:
            values['class'].append(f'bx--tooltip--{position}')


    def render_slot_icon(self, values, context):
        """Render html of the slot.
        """
        return modify_svg(values['child'], {
            'focusable': 'false',
            'preserveAspectRatio': 'xMidYMid meet',
            'fill': 'currentColor',
            'style': {
                'width': '%spx' % 16,
                'height': '%spx' % 16,
            },
            'aria-hidden': 'true',
        })


class Interactive(BaseTooltip):
    """Tooltip component.
    """
    SLOTS = ('icon', 'footer', 'heading')
    "Named children."
    MODES = ('interactive', 'nolabel')
    "Available variants."
    REQUIRED_PROPS = ('label',)
    "Will raise Exception if not set."
    CLASS_AND_PROPS = ('label', 'content')
    "Prepare xxx_class and xxx_props values."

    def prepare(self, values, context):
        """Prepare values for rendering the templates.
        """
        if 'heading' in self.slots:
            values['content_props'].append(
                    ('aria-labelledby', 'heading-' + self._id))
        elif self.mode == 'nolabel':
            values['content_props'].append(('aria-label',
                    values['label'] + values['label_suffix']))
        else:
            values['content_props'].append(
                    ('aria-labelledby', 'label-' + self._id))


    def render_interactive(self, values, context):
        """Output html of the component.
        """
        template = """
<div id="label-{id}" class="bx--tooltip__label">
  {label}{label_suffix}
  <button aria-expanded="false" aria-labelledby="label-{id}"
      data-tooltip-trigger data-tooltip-target="#{id}"
      class="bx--tooltip__trigger {class}" aria-controls="{id}" {props}>
    {slot_icon}
  </button>
</div>
<div id="{id}" aria-hidden="true" data-floating-menu-direction="bottom"
    class="bx--tooltip">
  <span class="bx--tooltip__caret"></span>
  <div class="bx--tooltip__content {content_class}" tabindex="-1" role="dialog"
      aria-describedby="body-{id}" {content_props}>
    {slot_heading}
    <p id="body-{id}">{child}</p>
    {slot_footer}
  </div>
  <span tabindex="0"></span>
</div>
"""
        return self.format(template, values, context)


    def render_nolabel(self, values, context):
        """Output html of the component.
        """
        template = """
<div id="label-{id}" class="bx--tooltip__label">
  {label}{label_suffix}
  <div tabindex="0" aria-expanded="false" aria-labelledby="label-{id}"
      data-tooltip-trigger data-tooltip-target="#{id}"
      role="button" class="bx--tooltip__trigger {class}" aria-controls="{id}" {props}>
    {slot_icon}
  </div>
</div>
<div id="{id}" aria-hidden="true" data-floating-menu-direction="bottom"
    class="bx--tooltip">
  <span class="bx--tooltip__caret"></span>
  <div class="bx--tooltip__content {content_class}" tabindex="-1" role="dialog"
      aria-describedby="body-{id}" {content_props}>
    {slot_heading}
    <p id="body-{id}">{child}</p>
    {slot_footer}
  </div>
  <span tabindex="0"></span>
</div>
"""
        return self.format(template, values, context)


    def render_slot_heading(self, values, context):
        """Render html of the slot.
        """
        template = """
<h4 id="heading-{id}" class="bx--tooltip__heading {class}" {props}>{child}</h4>
"""
        return self.format(template, values)


    def render_slot_footer(self, values, context):
        """Render html of the slot.
        """
        template = """
<div class="bx--tooltip__footer {class}" {props}>
  {child}
</div>
"""
        return self.format(template, values)


class Definition(BaseTooltip):
    """Tooltip component.
    """
    REQUIRED_PROPS = ('label',)
    "Will raise Exception if not set."

    def render_default(self, values, context):
        """Output html of the component.
        """
        template = """
<div class="bx--tooltip--definition bx--tooltip--a11y" data-tooltip-definition>
  <button aria-describedby="{id}"
      class="bx--tooltip__trigger bx--tooltip--a11y bx--tooltip__trigger--definition {class}"
      {props}>
    {label}{label_suffix}
  </button>
  <div class="bx--assistive-text" id="{id}" role="tooltip">
    {child}
  </div>
</div>
"""
        return self.format(template, values)


class Icon(BaseTooltip):
    """Tooltip component.
    """
    SLOTS = ('icon',)
    "Named children."

    def render_default(self, values, context):
        """Output html of the component.
        """
        template = """
<button class="bx--tooltip__trigger bx--tooltip--a11y {class}"
    data-tooltip-icon {props}>
  <span class="bx--assistive-text">{child}</span>
  {slot_icon}
</button>
"""
        return self.format(template, values, context)


components = {
    'InteractiveTooltip': Interactive,
    'DefinitionTooltip': Definition,
    'IconTooltip': Icon,
}
